### PR TITLE
Borders: Stabilize border block supports within block processing

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Part of: https://github.com/WordPress/gutenberg/issues/64312

Related:
- https://github.com/WordPress/gutenberg/pull/63401
- https://github.com/WordPress/gutenberg/pull/66849

## What?

This PR aims to stabilize the border block support key, i.e. `__experimentalBorder` --> `border`.

Props to @hbhalodia and @benazeer-ben for their work to date in this area (see https://github.com/WordPress/gutenberg/pull/64330 & https://github.com/WordPress/gutenberg/pull/65968 respectively)

## Why?

Border support has been around for a long time in its "experimental" form. Recent efforts to drive its adoption across core blocks highlight the need to make it stable. It's time.

Also, as the original issue notes:

> Doing so will make it easier for third-party extenders to confidently build custom blocks that provide border support and modify existing blocks with this support.

## How?

This PR builds upon the path to stabilizing block supports charted in https://github.com/WordPress/gutenberg/pull/63401.

- The border block support and Global Styles have been updated to use the stable `border` key
- A PHP filter has been added to run last and stabilize any use of the `__experimentalBorder` key
- The `processBlockType` function in the blocks store has been updated to stabilize the border key as per [typography supports](https://github.com/WordPress/gutenberg/pull/63401)
- As discussed on the [typography support stabilization PR](https://github.com/WordPress/gutenberg/pull/63401), it's unlikely we can cover all permutations of filters and plugins that might update their use of experimental support keys at different rates. The approach settled on in these PRs covers the vast majority of use cases.
- Additional unit tests have been added or updated to cover border support stabilization. These also help internally document the expected outcomes for the stable block support.

## Testing Instructions

<details>
<summary>Example block markup</summary>

The following snippet is a group block with background, border, and padding applied. It contains a paragraph and image block each with its own borders too.

```html
<!-- wp:group {"tagName":"main","style":{"spacing":{"margin":{"top":"var:preset|spacing|60"}}},"layout":{"type":"constrained"}} -->
<main class="wp-block-group" style="margin-top:var(--wp--preset--spacing--60)"><!-- wp:heading {"textAlign":"left","level":1} -->
<h1 class="wp-block-heading has-text-align-left">Blog</h1>
<!-- /wp:heading -->

<!-- wp:group {"style":{"border":{"radius":{"topRight":"50px","bottomLeft":"50px"},"color":"#f0a402","style":"solid","width":"5px"},"spacing":{"padding":{"top":"var:preset|spacing|40","right":"var:preset|spacing|40","bottom":"var:preset|spacing|40","left":"var:preset|spacing|40"}},"color":{"background":"#8b4513"}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group has-border-color has-background" style="border-color:#f0a402;border-style:solid;border-width:5px;border-top-right-radius:50px;border-bottom-left-radius:50px;background-color:#8b4513;padding-top:var(--wp--preset--spacing--40);padding-right:var(--wp--preset--spacing--40);padding-bottom:var(--wp--preset--spacing--40);padding-left:var(--wp--preset--spacing--40)"><!-- wp:paragraph {"style":{"border":{"color":"#D2691E","width":"2px","style":"dotted"},"spacing":{"padding":{"top":"10px","right":"10px","bottom":"10px","left":"10px"}}},"textColor":"base"} -->
<p class="has-border-color has-base-color has-text-color" style="border-color:#D2691E;border-style:dotted;border-width:2px;padding-top:10px;padding-right:10px;padding-bottom:10px;padding-left:10px">Exploring the vast landscapes of Africa brings a new sense of wonder.</p>
<!-- /wp:paragraph -->

<!-- wp:image {"sizeSlug":"large","linkDestination":"none","style":{"border":{"color":"#ffd700","style":"solid","width":"3px"}}} -->
<figure class="wp-block-image size-large has-custom-border"><img src="https://unsplash.com/photos/wdMWMHXUpsc/download?ixid=M3wxMjA3fDB8MXxzZWFyY2h8M3x8YWZyaWNhfGVufDB8fHx8MTczMTQ0MzAxN3ww&amp;force=true&amp;w=640" alt="African landscape" class="has-border-color" style="border-color:#ffd700;border-style:solid;border-width:3px"/></figure>
<!-- /wp:image --></div>
<!-- /wp:group -->

```
</details>

<details>
<summary>Diff to update Group block to use stable border support key</summary>

```diff
diff --git a/packages/block-library/src/group/block.json b/packages/block-library/src/group/block.json
index 3c7d8d3ce1..77e956ace7 100644
--- a/packages/block-library/src/group/block.json
+++ b/packages/block-library/src/group/block.json
@@ -58,7 +58,7 @@
 		"dimensions": {
 			"minHeight": true
 		},
-		"__experimentalBorder": {
+		"border": {
 			"color": true,
 			"radius": true,
 			"style": true,

```
</details>

<details>
<summary>Filter to test 3rd party use of old experimental key</summary>

```diff
function disable_group_borders( $args ) {
	if ( 'core/group' === $args['name'] ) {
		_wp_array_set( $args, array( 'supports', '__experimentalBorder' ), false );
	}

	return $args;
}
add_filter( 'register_block_type_args', 'disable_group_borders', 20 );
```
</details>

1. Add the above test markup to a post or template
2. Ensure the border displays correctly in the editor and the frontend
3. Check that existing blocks with border support via the `__experimentalBorder` key still display the border controls within the block inspector and Global Styles.
4. Confirm that blocks opting to make border controls display by default (e.g. Group), still do. Likewise, check blocks that hide border controls by default (e.g. Site Title), or only hide some border controls (e.g. Avatar) still behave too.
5. Apply the above diff to update the Group block to use the stable border support key, or simply edit its block.json file, changing `__experimentalBorder` to `border`.
6. Confirm that borders for Group blocks can still be configured in Global Styles or on an individual block instance.
7. Add the filter from the above snippet to your theme functions.php file. This filter still uses the experimental key to test some backwards compatibility.
8. The earlier patch updated Group to use the stable border support key, the filter in the step above attempts to disable border support. Despite the filter using the old experimental key, this should work. Confirm you can no longer adjust borders on Group blocks i.e border controls no longer appear (existing block support styles will still show).
9. Test applying border styles via Global Styles for the Image block works as expected in the editor and frontend.
10. Update the Image block's supports to use the stable `border` key. Confirm the global border styles for the Image block are still correct.

## Screenshots or screencast 

There should be no visual changes. Any borders that don't match trunk are likely a regression.

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
Quick note: I need to double-check and test a few more edge cases on this before it will be ready for review. I'll flag accordingly when it is.

**Comment by aaronrobertshaw** [comment]
Update:

~The stabilization of `__experimentalBorder` will also require stabilization within the [Block Selectors API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/). I'm currently looking into making this happen alongside `supports` stabilization.~

The stabilization of the `__experimentalBorder` support key appears to causing an issue with the [Block Selectors API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/). I'll keep this PR in draft status until it is sorted.

**Comment by aaronrobertshaw** [comment]
The issue [noted above](https://github.com/WordPress/gutenberg/pull/66918#issuecomment-2472028035) seems to have been isolated to how the global styles unit test set up the fake image block. Fixed in https://github.com/WordPress/gutenberg/pull/66918/commits/eb7585525328f81cbba8f131ec1f7613081917f1.

I've run through the test instructions in the PR description a couple of times in different ways. It's behaving pretty well for me. Fresh eyes might be able to see some edge cases I've missed though. I'll flag this ready for review now.

Given the backport will take a slightly different approach, I'll follow up with a backport once I'm sure the approach here is stable. A single backport might also cover both the typography stabilization and this.

_Note: The failing e2es are unrelated and should be temporary. I'll kick them off again in due course._

**Comment by andrewserong** [comment]
> The issue https://github.com/WordPress/gutenberg/pull/66918#issuecomment-2472028035 seems to have been isolated to how the global styles unit test set up the fake image block. Fixed in https://github.com/WordPress/gutenberg/commit/eb7585525328f81cbba8f131ec1f7613081917f1.

Ah, so in that case because it's passing `blockTypes` to `getBlockSelectors`, in real world usage, the experimental block support would have already been stabilized? Whereas in this case we need to update the test with the "already stabilized" version. That looks like a safe update to me 👍

File: packages/blocks/src/store/process-block-type.js
Comment (andrewserong) [inline packages/blocks/src/store/process-block-type.js]: Tiny nit: I like how the JS and PHP implementations match in terms of checking the order in which keys were added. The PHP implementation has slightly clearer comments. Would it be worth adding those two to the JS implementation? E.g.

```
// Determine the order of keys, so the last defined can be preferred.
```

And

```
// Update support config, prefer the last defined value.
```

My main thinking here is that it mightn't be clear to our future selves why we've gone with ensuring the most recently updated key wins out, so the extra comment might help with clarity. Not a big issue, though!

Comment (aaronrobertshaw) [reply packages/blocks/src/store/process-block-type.js]: Good idea, these bits were iterated on a little out of sync with each other, so they were due to be brought in line.

Comment (aaronrobertshaw) [reply packages/blocks/src/store/process-block-type.js]: Done: https://github.com/WordPress/gutenberg/pull/66918/commits/611de98df6a3345f6bb2b6810cde873b92d56b74

**Comment by aaronrobertshaw** [comment]
> Ah, so in that case because it's passing blockTypes to getBlockSelectors, in real world usage, the experimental block support would have already been stabilized?

That's correct. 

I'm open to ideas on better approaches. To me, the test here's aim was to cover the happy path of selector generation. The unit tests around stabilizing block supports cover ensuring that happens correctly.

**Comment by aaronrobertshaw** [comment]
Regarding the backport for this PR, after discussion with @andrewserong, the plan is to create a single backport for the stabilization of all block supports. That will happen via updates to the existing backport for typography supports in https://github.com/WordPress/wordpress-develop/pull/7069.

Some of the reasoning for this approach includes:
- The stabilization process is evolving with each block support and might change a little further when shared experimental props are stabilized i.e. `__experimentalDefaultControls` and `__experimentalSkipSerialization`.
- A single backport allows for all the stabiliization changes to be tested in unision in core. This should provide greater confidence in the updates and make any future debugging or historical digs much easier.

File: lib/compat/wordpress-6.8/blocks.php
Comment (andrewserong) [inline lib/compat/wordpress-6.8/blocks.php:69]: This is not at all a blocker for now, but in the JS implementation we check for an object, so there'd never be a match against the `align` support, where an array is a valid value. Here, there theoretically could be a match.

In practice it's not an issue because we don't have anything in `$experimental_to_stable_keys` that references supports where an array is a valid value, but just thought I'd mention it for the (very) unlikely event that at some distant point in the future we create an experimental support that does accept an array and then want to stabilize it further down the track. In that case, this condition might need to be updated.

For now, though, I think it's fine to leave as-is. And please excuse my verbose comment here 😄

**Review by andrewserong — APPROVED** [review]
Nice work here, I haven't been able to fault it! 🎉 

✅ The code changes look good to me, I like how it generalises the stabilization to include both typography and border, as well as creating a good structure for subsequent stabilization
✅ Thanks for improving the logic of the stabilization so that it factors in the order in which keys were added 👍 
✅ Works well as-is for experimental border and typography
✅ With the snippets applied, the stabilized keys appear to work well for static and dynamic blocks, in the post editor and in global styles
✅ Updating the image (and other) blocks to use the stabilized Border key works as expected in global styles

LGTM! 🚀 

We've chatted a bit about this work, so I've already got a fair amount of context for what's happening here. It might be worth getting a second review and/or opinion (e.g. possibly @ndiego) to confidence check as I'm potentially biased toward this approach, but I'm really liking this PR, along with the code quality improvements it adds!

File: lib/compat/wordpress-6.8/blocks.php
Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:69]: Appreciate you raising this one. It has vaguely been on my radar but I was struggling to come up with a real-word case where it should happen.

The `align` support is stable, so wouldn't be in the `$experimental_to_stable_keys` array, as you note. In that case, the guard clause at the beginning of the loop should just continue on, while using the `align` config as is.

In the end, rather than waste too much time on coming up with hypotheticals to test, I decided to leave this until it was a problem.

Does that allay your concern at all? What else am I missing?

Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:69]: I meant to also add that in recent times, there's been a decided effort to not land block supports at all unless they were stable. To prevent being in the situation like we are with some typography and border supports. That's been the case for entire block support features like `shadow` or subfeatures like `textColumns` and `minHeight`.

If the approach to not landing experimental block supports continues, there shouldn't be a scenario with a new experimental block support that needs stabilizing.

Comment (andrewserong) [reply lib/compat/wordpress-6.8/blocks.php:69]: Absolutely, what you've described captures my thinking, too, so you're not missing anything here!

> In the end, rather than waste too much time on coming up with hypotheticals to test, I decided to leave this until it was a problem.

That's where I was coming from, too. I was mostly thinking out loud, which is to say, flagging something that could _theoretically_ be an issue at some distant point in the future. And it's at that distant point (if we're adding an experimental support that uses arrays) where I think it could be addressed, rather than now 🙂

Comment (andrewserong) [reply lib/compat/wordpress-6.8/blocks.php:69]: > If the approach to not landing experimental block supports continues, there shouldn't be a scenario with a new experimental block support that needs stabilizing.

Indeed — my vote is to avoid using the `__experimental` prefix for things in the future, too 👍

Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php:69]: 👍 Having it on the record here will also flag it for our future selves or others, if it is really needed in the end.

**Comment by aaronrobertshaw** [comment]
This PR should be ready for a final second opinion now. 

There's a [second PR](https://github.com/WordPress/gutenberg/pull/67018) based off this branch that shows this stabilization approach can also work for the shared experimental support flags, `__experimentalSkipSerialization` and `__experimentalDefaultControls`.

The logic in the stabilization function changes significantly in the follow-up PR however I think there is a benefit to some focused testing for each iteration of stabilization. That and there are already a lot of files being touched here 🤷

**Comment by aaronrobertshaw** [comment]
@karthick-murugan, @hbhalodia, and @benazeer-ben given your experiences taking different approaches to stabilizing block supports, it would be great if you could help test this PR and https://github.com/WordPress/gutenberg/pull/67018 to ensure it's all working as expected.

I'm pretty confident this is the direction we wish to go. If there are no major issues, we can address any further edge cases in follow-ups so we can start getting this tested in the wild via the next Gutenberg release.

**Review by ramonjd — APPROVED** [review]
Not having much context on this, but I see @andrewserong has previously tests.

I went through all testing steps in TT5, and TT2 because I was feeling sentimental.

All the border supports work as expected with both the stabile and `__experimental` prefixes. Default controls appear as expected, e.g., Group vs Site Title. 

Setting `__experimentalBorder` to `false` via the filter disables the controls. 

Nothing in the error logs or console I can see. 

Nice job getting all the pieces together.

File: lib/compat/wordpress-6.8/blocks.php
Comment (ramonjd) [inline lib/compat/wordpress-6.8/blocks.php]: This could be just me, but should we add why the last defined is preferred? The logic looks good, but it code itself doesn't explain the reason.

I could be missing something, so feel free to ignore me. I'm only thinking out loud and wondering if folks looking at this code later might need the hint. 

> it's unlikely we can cover all permutations of filters and plugins that might update their use of experimental support keys at different rates.

Based on this, and what's been already explained to me, is it to increase the probability of accurately processing the last known support value, which could originate from late-running hooks?

Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php]: Yes. Filters could run in any order and we can't guarantee whether they'll use the experimental or stable key. 

This approach isn't bulletproof but should work for the vast majority of cases while becoming less and less of a possibility over time as extenders migrate to the stable key.

Some simple use cases this order-based approach tackles include:

1. Core blocks are updated to use stable keys but a plugin filters the block type args and sets a experimental key.
    - The expected outcome would be that the plugin's filtered value is use. 
    - If the block was already migrated to stable keys, the filtered experimental key would be added to the supports config after the stable key. 
    - Using the last defined key's value here allows the expected behaviour.
2. Take the flip side of the above scenario, the block uses an experimental key but a plugin filters things setting the stable key. 
    - Again as it was set later, then it should take precedence.

We get into trouble if multiple plugins all filter a blocks supports using a mix of experimental and stable keys one after another. There's no way we can control that scenario or determine with certainty which key's value should be used. The best we can do for now is use the insertion order for the keys. This is still better than simply using stable keys and only falling back to experimental keys in their absence.

Another option to mitigate a little of the risk here is to strongly recommend extenders update any block supports in their custom blocks or filters to use the stable keys ASAP. If a dev note isn't sufficient here, then a make post might be in order. 

All that aside, I'll try and improve that inline comment.

Comment (aaronrobertshaw) [reply lib/compat/wordpress-6.8/blocks.php]: I've had a crack at updating the inline comments in https://github.com/WordPress/gutenberg/pull/66918/commits/9c32c4ae41e6fb1bdae0044c622c9c2514f1f3fc. Let me know if you think that's adequate.

I'm happy to tweak further if there's a clearer, more eloquent, way of communicating this.

**Comment by karthick-murugan** [comment]
> @karthick-murugan, @hbhalodia, and @benazeer-ben given your experiences taking different approaches to stabilizing block supports, it would be great if you could help test this PR and #67018 to ensure it's all working as expected.
> 
> I'm pretty confident this is the direction we wish to go. If there are no major issues, we can address any further edge cases in follow-ups so we can start getting this tested in the wild via the next Gutenberg release.

@aaronrobertshaw - I have tested this PR with all the testing instructions above and all works perfectly. I have attached the video reference of the testing. Also tested by updating the functions.php with the code snippet and it works perfectly.

https://github.com/user-attachments/assets/450ec39a-c7ef-40ae-ab1d-61c925ab9ee2

**Comment by aaronrobertshaw** [comment]
Appreciate the review @karthick-murugan, thank you! 🙇 

I'll give Ramon a chance to see if the [updated inline comments](https://github.com/WordPress/gutenberg/pull/66918/commits/9c32c4ae41e6fb1bdae0044c622c9c2514f1f3fc) improve clarity. Then I'll aim to merge this tomorrow and push forward https://github.com/WordPress/gutenberg/pull/67018.

File: lib/compat/wordpress-6.8/blocks.php
Comment (ramonjd) [reply lib/compat/wordpress-6.8/blocks.php]: I appreciate the extra comments, thank you!

Don't let my questions stop the merge. Everything is iterable! 🚢

**Comment by aaronrobertshaw** [comment]
Quick update:

As discussed in https://github.com/WordPress/gutenberg/pull/67729#issuecomment-2553200639, the approach to block support stabilization needs more time so will likely be punted from 6.8. There is a revert PR available (https://github.com/WordPress/gutenberg/pull/68163) and anyone interested in stabilized block supports is advised to follow that for the latest 🙏

**Comment by ramonjd** [comment]
The work that's gone into this so far has been reverted in:

- https://github.com/WordPress/gutenberg/pull/68163

Thanks to everyone who helped move it forward, and tease out lessons for when folks return to it.

See background context: https://github.com/WordPress/gutenberg/pull/67729#issuecomment-2553200639

**Pull request merged**
